apollo-router/tests/integration/subgraph_response.rs

```diff
@@ -9,6 +9,78 @@ include_subgraph_errors:
   all: true
 "#;
 
+#[tokio::test(flavor = "multi_thread")]
+async fn test_subgraph_returning_data_null() -> Result<(), BoxError> {
+    let mut router = IntegrationTest::builder()
+        .config(CONFIG)
+        .responder(ResponseTemplate::new(200).set_body_json(json!({ "data": null })))
+        .build()
+        .await;
+
+    router.start().await;
+    router.assert_started().await;
+
+    let query = "{ __typename topProducts { name } }";
+    let (_trace_id, response) = router.execute_query(&json!({ "query":  query })).await;
+    assert_eq!(response.status(), 200);
+    assert_eq!(
+        serde_json::from_str::<serde_json::Value>(&response.text().await?)?,
+        json!({ "data": null })
+    );
+    Ok(())
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_subgraph_returning_different_typename_on_query_root() -> Result<(), BoxError> {
```

Comment:
are we sure in this test that the `__typename` selections got all the way to the subgraph? If that's the case, we are supposed to have some field rewriting driven by the query plan fetch nodes (`input_rewrites` and `output_rewrites`). Is that test exercising that rewriting, or what is happening in query formatting?

Reply:
Happening in query formatting, that's test that answer your comment here: https://github.com/apollographql/router/pull/6009#discussion_r1764961754

By switching to `apply_root_selection_set` we always execute this code (even if parts of root selection is wrapped in fragments): 
https://github.com/apollographql/router/blob/ddeed8d51df2c9c0edf19eb8fbb5530d31152295/apollo-router/src/spec/query.rs#L783-L786

So now we 100% sure __typename on root type is valid independently from anything that QP will return.

Reply:
that's good to make sure of that but probably we need a test that exercise the entire pipeline, not only the formatting

Reply:
yes, it's done in this test: https://github.com/apollographql/router/blob/e871c9331b9d1e7a66d646574f012f116f48b0c1/apollo-router/tests/integration/subgraph_response.rs#L34-L82

Reply:
it testing all possible scenario for how __typename can be nested and tests that __typename returned by subgraph are ignored.